---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  gene_symbol: "TP53"
  report_title: "Transciprts of `r params$gene_symbol`"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  report_folder: "../reports"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(EnsDb.Hsapiens.v86)
library(EnsDb.Mmusculus.v79)
library(ensembldb)
library(homologene)
library(GenomicRanges)
library(Gviz)
library(ggbio)
library(DT)

select <- dplyr::select
filter <- dplyr::filter

hsdb <- EnsDb.Hsapiens.v86
mmdb <- EnsDb.Mmusculus.v79
```

## Human transcripts

### Transcript annotations for gene

```{r}
Tx <- transcripts(hsdb, filter = GeneNameFilter(params$gene_symbol))
Tx <- Tx[Tx$tx_biotype != "LRG_gene"]
tx_df <- as.data.frame(Tx)

datatable(tx_df)
```

```{r}
autoplot(Tx)
```

```{r}
gene_models <- exonsBy(
  hsdb, by = "gene",
  filter = AnnotationFilterList(
    GeneNameFilter(params$gene_symbol),
    GeneIdFilter("ENS", "startsWith")
  )
)
autoplot(gene_models[[1]], aes(type = model))
```


## Mouse transcripts

### Transcript annotations for gene

```{r}
mouse_symbol <- homologene(params$gene_symbol, inTax = 9606, outTax = 10090)
Tx_m <- transcripts(mmdb, filter = GeneNameFilter(mouse_symbol))
txm_df <- as.data.frame(Tx_m)

datatable(txm_df)
```

