---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  gene_symbol: "TP53"
  report_title: "Transciprts of `r params$gene_symbol`"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  report_folder: "../reports"
  figure_folder: "../figures"
  table_folder: "../tables"
  data_folder: "../data"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(msa)
library(seqLogo)
library(EnsDb.Hsapiens.v86)
library(EnsDb.Mmusculus.v79)
library(ensembldb)
library(homologene)
library(GenomicRanges)
library(GenomicFeatures)
library(Biostrings)
library(Gviz)
library(ggbio)
library(ggseqlogo)
library(ggmsa)
library(bios2mds)
library(DT)

select <- dplyr::select
filter <- dplyr::filter

hsdb <- EnsDb.Hsapiens.v86
mmdb <- EnsDb.Mmusculus.v79
```

## Human transcripts

### Transcript annotations for gene

```{r}
Tx <- transcripts(
  hsdb, filter = GeneNameFilter(params$gene_symbol),
  columns = c("tx_biotype", "gene_name")#, "protein_id", "uniprot_id")
)
Tx <- Tx[Tx$tx_biotype != "LRG_gene"]
tx_df <- Tx %>%
  as.data.frame() %>%
  select(gene_name, chromosome=seqnames, start, end, tx_biotype)

datatable(tx_df)
```

```{r}
chr <- as.character(unique(seqnames(Tx)))
region_start = min(tx_df$start)
region_end = max(tx_df$end)

genome_location_track <- GenomeAxisTrack()
genome_region_range <- getGeneRegionTrackForGviz(
  hsdb, chromosome=chr, start=region_start, end=region_end,
  filter=AnnotationFilterList(
    GeneNameFilter(params$gene_symbol),
    GeneIdFilter("ENS", "startsWith")
  ), featureIs="tx_biotype"
)

genome_region_track <- GeneRegionTrack(
  genome_region_range, name="", transcriptAnnotation="transcript",
  background.title="white"
)

options(ucscChromosomeNames = FALSE)
plotTracks(list(genome_location_track, genome_region_track))
```

### Nucleotide sequences

```{r}
dna <- ensembldb:::getGenomeTwoBitFile(hsdb)

nt_seqs <- hsdb %>%
  exonsBy(
    by = "tx",
    filter = AnnotationFilterList(
      GeneNameFilter(params$gene_symbol),
      GeneIdFilter("ENS", "startsWith")
    )
  ) %>%
  extractTranscriptSeqs(dna, .)

nt_seqs
```

```{r}
if(!dir.exists(params$data_folder)) dir.create(params$data_folder, recursive=TRUE)

writeXStringSet(nt_seqs, file.path(params$data_folder, paste(params$gene_symbol, "_nt_human.fa", sep="_")))
```


### Protein sequences

```{r}
protein_seqs <- proteins(
  hsdb, return.type = "AAStringSet",
  filter = AnnotationFilterList(
    GeneNameFilter(params$gene_symbol),
    GeneIdFilter("ENS", "startsWith")
  )
)

writeXStringSet(protein_seqs, file.path(params$data_folder, paste(params$gene_symbol, "_aa_human.fa", sep="_")))

protein_seqs
```

```{r}
alignment_fn <- file.path(params$data_folder, paste(params$gene_symbol, "_msa_human.fa", sep="_"))

protein_seqs %>%
  msa() %>%
  msaConvert("bios2mds::align") %>%
  export.fasta(outfile=alignment_fn, ncol=60, open="w")

#msaPrettyPrint(aligned_prots, output="pdf", showNames="none", showLogo="none", askForOverwrite=FALSE, verbose=FALSE)
#print(aligned_prots, show="complete")

ggmsa(alignment_fn, start = 221, end = 280, char_width = 0.5, seq_name = T) +
  geom_seqlogo() +
  geom_msaBar()
```

```{r}
aligned_prots %>%
  consensusMatrix(as.prob=TRUE) %>%
  seqLogo()
```

```{r}
p <- ggmsa(protein_seqs, start = 221, end = 280, char_width = 0.5, seq_name = T) +
  geom_seqlogo() +
  geom_msaBar()

p
```

```{r}
p <- ggmsa(Biostrings::readAAStringSet(aligned_prots), start = 221, end = 280, char_width = 0.5, seq_name = T) +
  geom_seqlogo() +
  geom_msaBar()

p
```

```{r}
ggseqlogo(aligned_prots)
```


## Mouse transcripts

### Transcript annotations for gene

```{r}
mouse_symbol <- homologene(params$gene_symbol, inTax = 9606, outTax = 10090)
Tx_m <- transcripts(mmdb, filter = GeneNameFilter(mouse_symbol))
txm_df <- as.data.frame(Tx_m)

datatable(txm_df)
```

### Zebrafish

```{r}
danio_symbol <- homologene(params$gene_symbol, inTax = 9606, outTax = 7955)
bsg <- BSgenome.Drerio.UCSC.danRer10

extractTranscriptSeqs(bsg, gene_models)
```

### Drosophila

```{r}
drosi_symbol <- homologene(params$gene_symbol, inTax = 9606, outTax = 7227)
bsg <- BSgenome.Drerio.UCSC.danRer10

extractTranscriptSeqs(bsg, gene_models)
```

