---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  gene_symbol: "TP53"
  report_title: "Transciprts of `r params$gene_symbol`"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  figure_folder: "../figures"
  data_folder: "../data"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(msa)
library(seqLogo)
library(ensembldb)
library(homologene)
library(GenomicRanges)
library(GenomicFeatures)
library(Biostrings)
library(Gviz)
library(ggbio)
library(ggseqlogo)
library(ggmsa)
library(bios2mds)
library(DT)
library(httr)

select <- dplyr::select
filter <- dplyr::filter

for (dr in c('figure_folder', 'table_folder')){
  drn <- params[[dr]]
  if(!dir.exists(drn)) dir.create(drn, recursive=TRUE)
}
```

```{r}
library(EnsDb.Hsapiens.v86)
library(EnsDb.Mmusculus.v79)
library(TxDb.Drerio.UCSC.danRer10.refGene)
library(BSgenome.Drerio.UCSC.danRer10)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
library(BSgenome.Dmelanogaster.UCSC.dm6)

hsdb <- EnsDb.Hsapiens.v86
mmdb <- EnsDb.Mmusculus.v79
drtx <- TxDb.Drerio.UCSC.danRer10.refGene
drdb <- BSgenome.Drerio.UCSC.danRer10
dmtx <- TxDb.Dmelanogaster.UCSC.dm6.ensGene
dmdb <- BSgenome.Dmelanogaster.UCSC.dm6

hub_dir <- "/home/rstudio/data/TwoBit"
if(!dir.exists(hub_dir)) dir.create(hub_dir, recursive=TRUE)
AnnotationHub::setAnnotationHubOption("CACHE", hub_dir)
hs_dna <- ensembldb:::getGenomeTwoBitFile(hsdb)
mm_dna <- ensembldb:::getGenomeTwoBitFile(mmdb)

uniprot_base_url <- "https://www.uniprot.org/uploadlists/"
uniprot_api_url <- "www.ebi.ac.uk/proteins/api"

options(ucscChromosomeNames = FALSE)

parse_uniprot <- function(uni_acc){
  uniprot_entry <- uni_acc %>%
    file.path(uniprot_api_url, "proteins", .) %>%
    GET() %>%
    content()
  
  features <- uniprot_entry$features %>%
    purrr::map_dfr(as.data.frame) %>%
    mutate(
      ACCESSION = uni_acc,
      SEQ = uniprot_entry$sequence$sequence
    )
  if (is.null(features$description)) features$description <- NA
  
  select(features, ACCESSION, type, category, description, begin, end, SEQ)
}
```

## Human transcripts

### Transcript annotations for gene

```{r}
Tx <- transcripts(
  hsdb, filter = GeneNameFilter(params$gene_symbol),
  columns = c("tx_biotype", "gene_name")#, "protein_id", "uniprot_id")
)
Tx <- Tx[Tx$tx_biotype != "LRG_gene"]
tx_df <- Tx %>%
  as.data.frame() %>%
  select(gene_name, chromosome=seqnames, start, end, tx_biotype)

datatable(tx_df)
```

```{r}
chr <- as.character(unique(seqnames(Tx)))
region_start = min(tx_df$start)
region_end = max(tx_df$end)

genome_location_track <- GenomeAxisTrack()
genome_region_range <- getGeneRegionTrackForGviz(
  hsdb, chromosome=chr, start=region_start, end=region_end,
  filter=AnnotationFilterList(
    GeneNameFilter(params$gene_symbol),
    GeneIdFilter("ENS", "startsWith")
  ), featureIs="tx_biotype"
)

genome_region_track <- GeneRegionTrack(
  genome_region_range, name="", transcriptAnnotation="transcript",
  background.title="white"
)

pdf(
  file.path(params$figure_folder, paste(params$gene_symbol, "transcripts_human.pdf", sep="_")),
  width=7.2, height=3.6
)
plotTracks(
  list(genome_location_track, genome_region_track),
  main=paste("Aligned protein sequence variants of", params$gene_symbol)
)
graphics.off()
plotTracks(list(genome_location_track, genome_region_track))
```

### Nucleotide sequences

```{r}
nt_seqs <- hsdb %>%
  exonsBy(
    by = "tx",
    filter = AnnotationFilterList(
      GeneNameFilter(params$gene_symbol),
      GeneIdFilter("ENS", "startsWith")
    )
  ) %>%
  extractTranscriptSeqs(hs_dna, .)

writeXStringSet(nt_seqs, file.path(params$data_folder, paste(params$gene_symbol, "nt_human.fa", sep="_")))

nt_seqs
```

### Protein sequences

```{r}
protein_seqs <- proteins(
  hsdb, return.type = "AAStringSet",
  filter = AnnotationFilterList(
    GeneNameFilter(params$gene_symbol),
    GeneIdFilter("ENS", "startsWith")
  )
)

writeXStringSet(protein_seqs, file.path(params$data_folder, paste(params$gene_symbol, "aa_human.fa", sep="_")))

protein_seqs
```

```{r}
alignment_fn <- file.path(params$data_folder, paste(params$gene_symbol, "msa_ensembl_human.fa", sep="_"))

prot_msa_list <- protein_seqs %>%
  msa() %>%
  msaConvert("bios2mds::align")

export.fasta(prot_msa_list, outfile=alignment_fn)

alignment_fn %>%
  readLines() %>%
  gsub("NA$", "", .) %>%
  writeLines(alignment_fn)

#msaPrettyPrint(aligned_prots, output="pdf", showNames="none", showLogo="none", askForOverwrite=FALSE, verbose=FALSE)
#print(aligned_prots, show="complete")
prot_msa <- readAAStringSet(alignment_fn)

prot_msa
```

```{r message=FALSE, warning=FALSE}
align_start <- 1
align_end <- 100
#align_start <- NULL
#align_end <- NULL

p <- ggmsa(alignment_fn, start=align_start, end=align_end, seq_name=TRUE) +
  geom_seqlogo() +
  geom_msaBar()

print(p)
```

```{r}
prot_msa_list %>%
  purrr::map(function(x) gsub("-", " ", x[1:50])) %>%
  purrr::map(paste, collapse="") %>%
  {.[. != ""]} %>%
  unlist() %>%
  ggseqlogo()
```

```{r}
aligned_ranges <- prot_msa %>%
  as.list() %>%
  purrr::map_df(
    function(x, y) {
      x %>%
        stringr::str_locate_all('[A-Z]+') %>%
        data.frame() %>%
        mutate(
          seqname = "P0",
          strand = "+"
        )
    }, .id="transcript"
  )

aligned_region_track <- aligned_ranges %>%
  makeGRangesFromDataFrame(keep.extra.columns=TRUE) %>%
  GeneRegionTrack(
  name="", transcriptAnnotation="transcript",
  background.title="white"
)

genome_location_track <- GenomeAxisTrack()

pdf(
  file.path(params$figure_folder, paste(params$gene_symbol, "ensembl_proteins_human.pdf", sep="_")),
  width=7.2, height=3.6
)
plotTracks(
  list(genome_location_track, aligned_region_track),
  main=paste("Aligned protein sequence variants of", params$gene_symbol),
  cex.main=1
)
graphics.off()
plotTracks(list(aligned_region_track, genome_location_track))
```

### Annotated features

```{r warning=FALSE, message=FALSE}
uniprot_id_mapping <- uniprot_base_url %>% 
  GET(
    query = list(
      from = "ENSEMBL_PRO_ID", to = "ACC", format = "tab",
      query = paste(names(protein_seqs), collapse=" ")
    )
  ) %>%
  content() %>%
  read.csv(text=., sep="\t", stringsAsFactors=FALSE)

protein_features <- uniprot_id_mapping$To %>%
  purrr::map_dfr(parse_uniprot) %>%
  left_join(uniprot_id_mapping, by=c(ACCESSION="To")) %>%
  dplyr::rename(ID = From) %>%
  distinct()

protein_features %>%
  select(-SEQ) %>%
  datatable()
```

```{r}
protein_seqs <- protein_features %>%
  distinct(ACCESSION, SEQ) %>%
  {setNames(.$SEQ, .$ACCESSION)} %>%
  AAStringSet()

writeXStringSet(protein_seqs, file.path(params$data_folder, paste(params$gene_symbol, "aa_human.fa", sep="_")))

protein_seqs
```

```{r}
alignment_fn <- file.path(params$data_folder, paste(params$gene_symbol, "msa_human.fa", sep="_"))

prot_msa_list <- protein_seqs %>%
  msa() %>%
  msaConvert("bios2mds::align")

export.fasta(prot_msa_list, outfile=alignment_fn)

alignment_fn %>%
  readLines() %>%
  gsub("NA$", "", .) %>%
  writeLines(alignment_fn)

prot_msa <- readAAStringSet(alignment_fn)

prot_msa
```

```{r}
msa_position_map <- aligned_ranges %>%
  dplyr::rename(ACCESSION = transcript) %>%
  mutate(
    Aligned_pos = purrr::map2(start, end, seq)
  ) %>%
  unnest(Aligned_pos) %>%
  arrange(ACCESSION, Aligned_pos) %>%
  group_by(ACCESSION) %>%
  mutate(
    Original_pos = seq(1, n())
  ) %>%
  ungroup() %>%
  select(ACCESSION, Original_pos, Aligned_pos)

msa_position_map
```

```{r}
aligned_ranges <- prot_msa %>%
  as.list() %>%
  purrr::map_df(
    function(x, y) {
      x %>%
        stringr::str_locate_all('[A-Z]+') %>%
        data.frame() %>%
        mutate(
          seqname = "P0",
          strand = "+"
        )
    }, .id="transcript"
  )

aligned_region_track <- aligned_ranges %>%
  makeGRangesFromDataFrame(keep.extra.columns=TRUE) %>%
  GeneRegionTrack(
  name="", transcriptAnnotation="transcript",
  background.title="white"
)

genome_location_track <- GenomeAxisTrack()

pdf(
  file.path(params$figure_folder, paste(params$gene_symbol, "ensembl_proteins_human.pdf", sep="_")),
  width=7.2, height=3.6
)
plotTracks(
  list(genome_location_track, aligned_region_track),
  main=paste("Aligned protein sequence variants of", params$gene_symbol),
  cex.main=1
)
graphics.off()
plotTracks(list(aligned_region_track, genome_location_track))
```

```{r}
main_prots <- c("P04637", "E7EQX7")

protein_features %>%
  filter(ACCESSION %in% main_prots) %>%
  filter(category == "PTM") %>%
  distinct(ACCESSION, begin, end, description) %>%
  arrange(description) %>%
  mutate(
    begin = as.numeric(begin),
    end = as.numeric(end),
    width = end-begin
  ) %>%
  left_join(msa_position_map, by=c("ACCESSION", begin="Original_pos")) %>%
  mutate(
    begin = Aligned_pos,
    end = begin + width
  ) %>%
  ggplot(aes(x=begin, y=description)) +
  geom_tile()
```



## Mouse transcripts

### Transcript annotations for gene

```{r}
mouse_symbol <- homologene(params$gene_symbol, inTax = 9606, outTax = 10090)$`10090`
Tx_m <- transcripts(mmdb, filter = GeneNameFilter(mouse_symbol))

Tx_m <- transcripts(
  mmdb, filter = GeneNameFilter(mouse_symbol),
  columns = c("tx_biotype", "gene_name")#, "protein_id", "uniprot_id")
)
Tx_m <- Tx_m[Tx_m$tx_biotype != "LRG_gene"]
txm_df <- Tx_m %>%
  as.data.frame() %>%
  select(gene_name, chromosome=seqnames, start, end, tx_biotype)

datatable(txm_df)
```

```{r}
chr <- as.character(unique(seqnames(Tx_m)))
region_start = min(txm_df$start)
region_end = max(txm_df$end)

genome_location_track <- GenomeAxisTrack()
genome_region_range <- getGeneRegionTrackForGviz(
  mmdb, chromosome=chr, start=region_start, end=region_end,
  filter=AnnotationFilterList(
    GeneNameFilter(mouse_symbol),
    GeneIdFilter("ENS", "startsWith")
  ), featureIs="tx_biotype"
)

genome_region_track <- GeneRegionTrack(
  genome_region_range, name="", transcriptAnnotation="transcript",
  background.title="white"
)

pdf(file.path(params$figure_folder, paste(params$gene_symbol, "transcripts_mouse.pdf", sep="_")))
plotTracks(list(genome_location_track, genome_region_track), width=7.2, height=3.6)
graphics.off()
plotTracks(list(genome_location_track, genome_region_track))
```

### Nucleotide sequences

```{r}
nt_seqs <- mmdb %>%
  exonsBy(
    by = "tx",
    filter = AnnotationFilterList(
      GeneNameFilter(mouse_symbol),
      GeneIdFilter("ENS", "startsWith")
    )
  ) %>%
  extractTranscriptSeqs(mm_dna, .)

writeXStringSet(nt_seqs, file.path(params$data_folder, paste(params$gene_symbol, "nt_mouse.fa", sep="_")))

nt_seqs
```

### Protein sequences

```{r}
protein_seqs <- proteins(
  mmdb, return.type = "AAStringSet",
  filter = AnnotationFilterList(
    GeneNameFilter(mouse_symbol),
    GeneIdFilter("ENS", "startsWith")
  )
)

writeXStringSet(protein_seqs, file.path(params$data_folder, paste(params$gene_symbol, "aa_mouse.fa", sep="_")))

protein_seqs
```

```{r warning=FALSE, message=FALSE}
uniprot_id_mapping <- uniprot_base_url %>% 
  GET(
    query = list(
      from = "ENSEMBL_PRO_ID", to = "ACC", format = "tab",
      query = paste(names(protein_seqs), collapse=" ")
    )
  ) %>%
  content() %>%
  read.csv(text=., sep="\t", stringsAsFactors=FALSE)

protein_features <- uniprot_id_mapping$To %>%
  purrr::map_dfr(parse_uniprot)

protein_features %>%
  select(-SEQ) %>%
  datatable()
```

## Zebrafish

### Transcript annotations for gene

```{r}
danio_id <- params$gene_symbol %>%
  homologene(inTax = 9606, outTax = 7955) %>%
  .[["7955_ID"]] %>%
  as.character()

Tx_f <- transcriptsBy(drtx, "gene")[[danio_id]]
txf_df <- Tx_f %>%
  as.data.frame() %>%
  select(tx_id, tx_name, chromosome=seqnames, start, end)

datatable(txf_df)
```

```{r warning=FALSE, message=FALSE}
chr <- as.character(unique(seqnames(Tx_f)))
region_start = min(txf_df$start)
region_end = max(txf_df$end)

genome_location_track <- GenomeAxisTrack()
genome_region_range <- drtx %>%
  exonsBy("tx", use.names=TRUE) %>%
  .[unique(txf_df$tx_name)] %>%
  unlist()
elementMetadata(genome_region_range)$transcript <- names(genome_region_range)

genome_region_track <- GeneRegionTrack(
  genome_region_range, name="", transcriptAnnotation="transcript",
  background.title="white"
)

pdf(file.path(params$figure_folder, paste(params$gene_symbol, "transcripts_fish.pdf", sep="_")))
plotTracks(list(genome_location_track, genome_region_track), width=7.2, height=3.6)
graphics.off()
plotTracks(list(genome_location_track, genome_region_track))
```

### Nucleotide sequences

```{r}
nt_seqs <- drdb %>%
  getSeq(Tx_f) %>%
  setNames(txf_df$tx_name)

writeXStringSet(nt_seqs, file.path(params$data_folder, paste(params$gene_symbol, "nt_fish.fa", sep="_")))

nt_seqs
```

### Protein sequences

```{r warning=FALSE, message=FALSE}
uniprot_id_mapping <- uniprot_base_url %>% 
  GET(
    query = list(
      from = "REFSEQ_NT_ID", to = "ACC", format = "tab",
      query = paste(txf_df$tx_name, collapse=" ")
    )
  ) %>%
  content() %>%
  read.csv(text=., sep="\t", stringsAsFactors=FALSE)

parse_uniprot <- function(uni_acc){
  uniprot_entry <- uni_acc %>%
    file.path(uniprot_api_url, "proteins", .) %>%
    GET() %>%
    content()
  
  uniprot_entry$features %>%
    purrr::map_dfr(as.data.frame) %>%
    mutate(
      ACCESSION = uni_acc,
      SEQ = uniprot_entry$sequence$sequence
    )
}

protein_features <- uniprot_id_mapping$To %>%
  purrr::map_dfr(parse_uniprot)

protein_features %>%
  select(-SEQ) %>%
  datatable()
```

```{r}
protein_seqs <- protein_features %>%
  distinct(ACCESSION, SEQ) %>%
  {setNames(.$SEQ, .$ACCESSION)} %>%
  AAStringSet(protein_seqs)

writeXStringSet(protein_seqs, file.path(params$data_folder, paste(params$gene_symbol, "aa_fish.fa", sep="_")))

protein_seqs
```

## Drosophila

### Transcript annotations for gene

```{r}
drpsi_id <- params$gene_symbol %>%
  homologene(inTax = 9606, outTax = 7227) %>%
  .[["7227_ID"]] %>%
  as.character()

Tx_d <- transcriptsBy(dmtx, "gene")[[drosi_id]]
txd_df <- Tx_d %>%
  as.data.frame() %>%
  select(tx_id, tx_name, chromosome=seqnames, start, end)

datatable(txd_df)
```

```{r warning=FALSE, message=FALSE}
chr <- as.character(unique(seqnames(Tx_d)))
region_start = min(txd_df$start)
region_end = max(txd_df$end)

genome_location_track <- GenomeAxisTrack()
genome_region_range <- dmtx %>%
  exonsBy("tx", use.names=TRUE) %>%
  .[unique(txd_df$tx_name)] %>%
  unlist()
elementMetadata(genome_region_range)$transcript <- names(genome_region_range)

genome_region_track <- GeneRegionTrack(
  genome_region_range, name="", transcriptAnnotation="transcript",
  background.title="white"
)

pdf(file.path(params$figure_folder, paste(params$gene_symbol, "transcripts_fly.pdf", sep="_")))
plotTracks(list(genome_location_track, genome_region_track), width=7.2, height=3.6)
graphics.off()
plotTracks(list(genome_location_track, genome_region_track))
```

### Nucleotide sequences

```{r}
nt_seqs <- dmdb %>%
  getSeq(Tx_d) %>%
  setNames(txd_df$tx_name)

writeXStringSet(nt_seqs, file.path(params$data_folder, paste(params$gene_symbol, "nt_fly.fa", sep="_")))

nt_seqs
```

### Protein sequences

```{r warning=FALSE, message=FALSE}
uniprot_id_mapping <- uniprot_base_url %>% 
  GET(
    query = list(
      from = "REFSEQ_NT_ID", to = "ACC", format = "tab",
      query = paste(txf_df$tx_name, collapse=" ")
    )
  ) %>%
  content() %>%
  read.csv(text=., sep="\t", stringsAsFactors=FALSE)

protein_features <- uniprot_id_mapping$To %>%
  purrr::map_dfr(parse_uniprot)

protein_features %>%
  select(-SEQ) %>%
  datatable()
```

```{r}
protein_seqs <- protein_features %>%
  distinct(ACCESSION, SEQ) %>%
  {setNames(.$SEQ, .$ACCESSION)} %>%
  AAStringSet(protein_seqs)

writeXStringSet(protein_seqs, file.path(params$data_folder, paste(params$gene_symbol, "aa_fish.fa", sep="_")))

protein_seqs
```
