---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  gene_symbol: "TP53"
  report_title: "Transciprts of `r params$gene_symbol`"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  report_folder: "../reports"
  figure_folder: "../figures"
  table_folder: "../tables"
  data_folder: "../data"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(msa)
library(seqLogo)
library(ensembldb)
library(homologene)
library(GenomicRanges)
library(GenomicFeatures)
library(Biostrings)
library(Gviz)
library(ggbio)
library(ggseqlogo)
library(ggmsa)
library(bios2mds)
library(DT)

select <- dplyr::select
filter <- dplyr::filter

for (dr in c('report_folder', 'figure_folder', 'table_folder', 'data_folder')){
  drn <- params[[dr]]
  if(!dir.exists(drn)) dir.create(drn, recursive=TRUE)
}
```

```{r}
library(EnsDb.Hsapiens.v86)
library(EnsDb.Mmusculus.v79)
library(TxDb.Drerio.UCSC.danRer10.refGene)
library(BSgenome.Drerio.UCSC.danRer10)
library(BSgenome.Dmelanogaster.UCSC.dm6)

hsdb <- EnsDb.Hsapiens.v86
mmdb <- EnsDb.Mmusculus.v79
drtx <- TxDb.Drerio.UCSC.danRer10.refGene
drdb <- BSgenome.Drerio.UCSC.danRer10
dmdb <- BSgenome.Dmelanogaster.UCSC.dm6

hub_dir <- "work/hub"
if(!dir.exists(hub_dir)) dir.create(hub_dir, recursive=TRUE)
setAnnotationHubOption("CACHE", hub_dir)
hs_dna <- ensembldb:::getGenomeTwoBitFile(hsdb)
mm_dna <- ensembldb:::getGenomeTwoBitFile(mmdb)
```

## Human transcripts

### Transcript annotations for gene

```{r}
Tx <- transcripts(
  hsdb, filter = GeneNameFilter(params$gene_symbol),
  columns = c("tx_biotype", "gene_name")#, "protein_id", "uniprot_id")
)
Tx <- Tx[Tx$tx_biotype != "LRG_gene"]
tx_df <- Tx %>%
  as.data.frame() %>%
  select(gene_name, chromosome=seqnames, start, end, tx_biotype)

datatable(tx_df)
```

```{r}
chr <- as.character(unique(seqnames(Tx)))
region_start = min(tx_df$start)
region_end = max(tx_df$end)

genome_location_track <- GenomeAxisTrack()
genome_region_range <- getGeneRegionTrackForGviz(
  hsdb, chromosome=chr, start=region_start, end=region_end,
  filter=AnnotationFilterList(
    GeneNameFilter(params$gene_symbol),
    GeneIdFilter("ENS", "startsWith")
  ), featureIs="tx_biotype"
)

genome_region_track <- GeneRegionTrack(
  genome_region_range, name="", transcriptAnnotation="transcript",
  background.title="white"
)

options(ucscChromosomeNames = FALSE)
plotTracks(list(genome_location_track, genome_region_track))
```

### Nucleotide sequences

```{r}
nt_seqs <- hsdb %>%
  exonsBy(
    by = "tx",
    filter = AnnotationFilterList(
      GeneNameFilter(params$gene_symbol),
      GeneIdFilter("ENS", "startsWith")
    )
  ) %>%
  extractTranscriptSeqs(hs_dna, .)

writeXStringSet(nt_seqs, file.path(params$data_folder, paste(params$gene_symbol, "nt_human.fa", sep="_")))

nt_seqs
```

### Protein sequences

```{r}
protein_seqs <- proteins(
  hsdb, return.type = "AAStringSet",
  filter = AnnotationFilterList(
    GeneNameFilter(params$gene_symbol),
    GeneIdFilter("ENS", "startsWith")
  )
)

writeXStringSet(protein_seqs, file.path(params$data_folder, paste(params$gene_symbol, "aa_human.fa", sep="_")))

protein_seqs
```

```{r}
alignment_fn <- file.path(params$data_folder, paste(params$gene_symbol, "msa_human.fa", sep="_"))

prot_msa_list <- protein_seqs %>%
  msa() %>%
  msaConvert("bios2mds::align")

export.fasta(prot_msa_list, outfile=alignment_fn)

alignment_fn %>%
  readLines() %>%
  gsub("NA$", "", .) %>%
  writeLines(alignment_fn)

#msaPrettyPrint(aligned_prots, output="pdf", showNames="none", showLogo="none", askForOverwrite=FALSE, verbose=FALSE)
#print(aligned_prots, show="complete")
prot_msa <- readAAStringSet(alignment_fn)

prot_msa
```

```{r message=FALSE, warning=FALSE}
align_start <- 1
align_end <- 100
#align_start <- NULL
#align_end <- NULL

p <- ggmsa(alignment_fn, start=align_start, end=align_end, seq_name=TRUE) +
  geom_seqlogo() +
  geom_msaBar()

print(p)
```

```{r}
prot_msa_list %>%
  purrr::map(function(x) gsub("-", " ", x[1:50])) %>%
  purrr::map(paste, collapse="") %>%
  {.[. != ""]} %>%
  unlist() %>%
  ggseqlogo()
```

## Mouse transcripts

### Transcript annotations for gene

```{r}
mouse_symbol <- homologene(params$gene_symbol, inTax = 9606, outTax = 10090)$`10090`
Tx_m <- transcripts(mmdb, filter = GeneNameFilter(mouse_symbol))

Tx_m <- transcripts(
  mmdb, filter = GeneNameFilter(mouse_symbol),
  columns = c("tx_biotype", "gene_name")#, "protein_id", "uniprot_id")
)
Tx_m <- Tx_m[Tx_m$tx_biotype != "LRG_gene"]
txm_df <- Tx_m %>%
  as.data.frame() %>%
  select(gene_name, chromosome=seqnames, start, end, tx_biotype)

datatable(txm_df)
```

```{r}
chr <- as.character(unique(seqnames(Tx_m)))
region_start = min(txm_df$start)
region_end = max(txm_df$end)

genome_location_track <- GenomeAxisTrack()
genome_region_range <- getGeneRegionTrackForGviz(
  mmdb, chromosome=chr, start=region_start, end=region_end,
  filter=AnnotationFilterList(
    GeneNameFilter(mouse_symbol),
    GeneIdFilter("ENS", "startsWith")
  ), featureIs="tx_biotype"
)

genome_region_track <- GeneRegionTrack(
  genome_region_range, name="", transcriptAnnotation="transcript",
  background.title="white"
)

options(ucscChromosomeNames = FALSE)
plotTracks(list(genome_location_track, genome_region_track))
```

### Nucleotide sequences

```{r}
nt_seqs <- mmdb %>%
  exonsBy(
    by = "tx",
    filter = AnnotationFilterList(
      GeneNameFilter(mouse_symbol),
      GeneIdFilter("ENS", "startsWith")
    )
  ) %>%
  extractTranscriptSeqs(mm_dna, .)

writeXStringSet(nt_seqs, file.path(params$data_folder, paste(params$gene_symbol, "nt_mouse.fa", sep="_")))

nt_seqs
```

### Protein sequences

```{r}
protein_seqs <- proteins(
  mmdb, return.type = "AAStringSet",
  filter = AnnotationFilterList(
    GeneNameFilter(mouse_symbol),
    GeneIdFilter("ENS", "startsWith")
  )
)

writeXStringSet(protein_seqs, file.path(params$data_folder, paste(params$gene_symbol, "aa_mouse.fa", sep="_")))

protein_seqs
```

## Zebrafish

```{r}
danio_id <- params$gene_symbol %>%
  homologene(inTax = 9606, outTax = 7955) %>%
  .[["7955_ID"]] %>%
  as.character()

Tx_f <- transcriptsBy(drtx, "gene")[[danio_id]]
txf_df <- Tx_f %>%
  as.data.frame() %>%
  select(tx_id, tx_name, chromosome=seqnames, start, end)

datatable(txf_df)
```

```{r}
gr <- exonsBy(drtx, by = "tx", use.names=TRUE)[unique(txf_df$tx_name)]
gr <- unlist(gr)
elementMetadata(gr)$transcript <- names(gr)
track <- Gviz::GeneRegionTrack(gr)
Gviz::plotTracks(track)
```

```{r warning=FALSE, message=FALSE}
chr <- as.character(unique(seqnames(Tx_f)))
region_start = min(txf_df$start)
region_end = max(txf_df$end)

genome_location_track <- GenomeAxisTrack()
genome_region_range <- drtx %>%
  exonsBy("tx", use.names=TRUE) %>%
  .[unique(txf_df$tx_name)] %>%
  unlist()
elementMetadata(genome_region_range)$transcript <- names(genome_region_range)

genome_region_track <- GeneRegionTrack(
  genome_region_range, name="", transcriptAnnotation="transcript",
  background.title="white"
)

plotTracks(list(genome_location_track, genome_region_track))
```

### Nucleotide sequences

```{r}
nt_seqs <- drdb %>%
  getSeq(Tx_f) %>%
  setNames(txf_df$tx_name)

writeXStringSet(nt_seqs, file.path(params$data_folder, paste(params$gene_symbol, "nt_mouse.fa", sep="_")))

nt_seqs
```

### Protein sequences

```{r}
# It is not straightforward to lookup protein sequences in available sources. Maybe Uniprot?
```

### Drosophila

```{r}
drosi_symbol <- homologene(params$gene_symbol, inTax = 9606, outTax = 7227)
dmdb <- BSgenome.Drerio.UCSC.danRer10
dmtx <- TxDb.Dmelanogaster.UCSC.dm6.ensGene

extractTranscriptSeqs(bsg, gene_models)
```

