---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  gene_symbol: "TP53"
  fly_symbol: "FBgn0039044"
  report_title: "Transcripts of `r params$gene_symbol`"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  figure_folder: "../figures"
  data_folder: "../data"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(msa)
library(ensembldb)
library(homologene)
library(GenomicRanges)
library(GenomicFeatures)
library(Biostrings)
library(Gviz)
library(ggplot2)
library(ggmsa)
library(bios2mds)
library(DT)
library(httr)

select <- dplyr::select
filter <- dplyr::filter

for (dr in c('figure_folder', 'data_folder')){
  drn <- params[[dr]]
  if(!dir.exists(drn)) dir.create(drn, recursive=TRUE)
}
```

```{r}
library(EnsDb.Hsapiens.v86)
library(EnsDb.Mmusculus.v79)
library(TxDb.Drerio.UCSC.danRer10.refGene)
library(BSgenome.Drerio.UCSC.danRer10)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
library(BSgenome.Dmelanogaster.UCSC.dm6)

hsdb <- EnsDb.Hsapiens.v86
mmdb <- EnsDb.Mmusculus.v79
drtx <- TxDb.Drerio.UCSC.danRer10.refGene
drdb <- BSgenome.Drerio.UCSC.danRer10
dmtx <- TxDb.Dmelanogaster.UCSC.dm6.ensGene
dmdb <- BSgenome.Dmelanogaster.UCSC.dm6

hub_dir <- "/home/rstudio/data/TwoBit"
if(!dir.exists(hub_dir)) dir.create(hub_dir, recursive=TRUE)
AnnotationHub::setAnnotationHubOption("CACHE", hub_dir)
hs_dna <- ensembldb:::getGenomeTwoBitFile(hsdb)
mm_dna <- ensembldb:::getGenomeTwoBitFile(mmdb)

source("../scripts/R/sequence_retrieval.R")
source("../scripts/R/alignment_blocks.R")
source("../scripts/R/transcript_viz.R")
```

## Human transcripts

### Transcript annotations for gene

```{r}
Tx <- transcripts(
  hsdb, filter = GeneNameFilter(params$gene_symbol),
  columns = c("tx_biotype", "gene_name")#, "protein_id", "uniprot_id")
)
Tx <- Tx[Tx$tx_biotype != "LRG_gene"]
tx_df <- Tx %>%
  as.data.frame() %>%
  select(gene_name, chromosome=seqnames, start, end, tx_biotype)

datatable(tx_df)
```

```{r}
fig_path = file.path(
  params$figure_folder,
  paste(params$gene_symbol, "transcripts_human.pdf", sep="_")
)
plot_trancript_models(params$gene_symbol, Tx, hsdb, fig_path)
```
Red regions are annotated as UTR, green as protein coding and blue as nonsense-mediated decay.

### Nucleotide sequences

```{r}
fasta_path = file.path(
  params$data_folder,
  paste(params$gene_symbol, "nt_human.fa", sep="_")
)
nt_seqs <- dna_seq_from_ensembl(params$gene_symbol, hsdb, hs_dna, fasta_path)
nt_seqs
```

### Protein sequences

```{r}
fasta_path = file.path(
  params$data_folder,
  paste(params$gene_symbol, "aa_ensembl_human.fa", sep="_")
)
protein_seqs <- aa_seq_from_ensembl(params$gene_symbol, hsdb, fasta_path)
protein_seqs
```

```{r}
alignment_fn <- file.path(
  params$data_folder,
  paste(params$gene_symbol,"msa_ensembl_human.fa", sep="_")
)
prot_msa <- align_isoforms(protein_seqs, alignment_fn)

prot_msa
```

```{r}
aligned_ranges <- convert_alignment_to_ranges(prot_msa)

plot_alignment_models(
  params$gene_symbol, aligned_ranges,
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "ensembl_proteins_human.pdf", sep="_")
  )
)
```

### Annotated features

```{r warning=FALSE, message=FALSE}
protein_features <- protein_seqs %>%
  names() %>%
  entries_from_uniprot()

protein_features %>%
  select(-SEQ) %>%
  datatable()
```

```{r}
fasta_path = file.path(
  params$data_folder,
  paste(params$gene_symbol, "aa_uniprot_human.fa", sep="_")
)

human_up_seqs <- extract_up_seqs(protein_features, fasta_path)

human_up_seqs
```

```{r}
alignment_fn <- file.path(
  params$data_folder,
  paste(params$gene_symbol, "msa_uniprot_human.fa", sep="_")
)
human_alignment <- align_isoforms(human_up_seqs, alignment_fn)

human_alignment
```

```{r}
aligned_ranges <- convert_alignment_to_ranges(human_alignment)
msa_position_map <- map_aligned_positions(aligned_ranges)

plot_alignment_models(
  params$gene_symbol, aligned_ranges,
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "uniprot_proteins_human.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "PTM", msa_position_map, paste("human", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "ptms_human.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "DOMAINS_AND_SITES", msa_position_map, paste("human", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "domains_human.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "STRUCTURAL", msa_position_map, paste("human", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "structure_human.pdf", sep="_")
  ), use_type=TRUE
)
```

## Mouse transcripts

### Transcript annotations for gene

```{r}
mouse_symbol <- homologene(params$gene_symbol, inTax = 9606, outTax = 10090)$`10090`

Tx_m <- transcripts(
  mmdb, filter = GeneNameFilter(mouse_symbol),
  columns = c("tx_biotype", "gene_name")
)
Tx_m <- Tx_m[Tx_m$tx_biotype != "LRG_gene"]
txm_df <- Tx_m %>%
  as.data.frame() %>%
  select(gene_name, chromosome=seqnames, start, end, tx_biotype)

datatable(txm_df)
```

```{r}
fig_path = file.path(
  params$figure_folder,
  paste(params$gene_symbol, "transcripts_mouse.pdf", sep="_")
)
plot_trancript_models(mouse_symbol, Tx_m, mmdb, fig_path)
```
Red regions are annotated as UTR, green as protein coding and blue as nonsense-mediated decay.

### Nucleotide sequences

```{r}
fasta_path = file.path(
  params$data_folder,
  paste(params$gene_symbol, "nt_mouse.fa", sep="_")
)
nt_seqs <- dna_seq_from_ensembl(mouse_symbol, mmdb, mm_dna, fasta_path)
nt_seqs
```

### Protein sequences

```{r warning=FALSE, message=FALSE}
protein_features <- nt_seqs %>%
  names() %>%
  entries_from_uniprot(id_code="ENSEMBL_TRS_ID")

protein_features %>%
  select(-SEQ) %>%
  datatable()
```

```{r}
fasta_path = file.path(
  params$data_folder,
  paste(params$gene_symbol, "aa_uniprot_mouse.fa", sep="_")
)

mouse_up_seqs <- extract_up_seqs(protein_features, fasta_path)

mouse_up_seqs
```

```{r}
alignment_fn <- file.path(
  params$data_folder,
  paste(params$gene_symbol, "msa_uniprot_mouse.fa", sep="_")
)
mouse_alignment <- align_isoforms(mouse_up_seqs, alignment_fn)

mouse_alignment
```

```{r}
aligned_ranges <- convert_alignment_to_ranges(mouse_alignment)
msa_position_map <- map_aligned_positions(aligned_ranges)

plot_alignment_models(
  mouse_symbol, aligned_ranges,
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "uniprot_proteins_mouse.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "PTM", msa_position_map, paste("mouse", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "ptms_mouse.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "DOMAINS_AND_SITES", msa_position_map, paste("mouse", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "domains_mouse.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "STRUCTURAL", msa_position_map, paste("mouse", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "structure_mouse.pdf", sep="_")
  ), use_type=TRUE
)
```

## Zebrafish

### Transcript annotations for gene

```{r}
fish_symbol <- params$gene_symbol %>%
  homologene(inTax = 9606, outTax = 7955) %>%
  .[["7955_ID"]] %>%
  as.character()

Tx_f <- transcriptsBy(drtx, "gene")[[fish_symbol]]
txf_df <- Tx_f %>%
  as.data.frame() %>%
  select(tx_id, tx_name, chromosome=seqnames, start, end)

datatable(txf_df)
```

```{r}
fig_path = file.path(
  params$figure_folder,
  paste(params$gene_symbol, "transcripts_fish.pdf", sep="_")
)
plot_trancript_models(fish_symbol, Tx_f, drtx, fig_path)
```

### Nucleotide sequences

```{r}
fasta_path = file.path(
  params$data_folder,
  paste(params$gene_symbol, "nt_fish.fa", sep="_")
)
nt_seqs <- dna_seq_from_txdb(Tx_f, drdb, fasta_path)
nt_seqs
```

### Protein sequences

```{r warning=FALSE, message=FALSE}
protein_features <- nt_seqs %>%
  names() %>%
  entries_from_uniprot(id_code="REFSEQ_NT_ID")

protein_features %>%
  select(-SEQ) %>%
  datatable()
```

```{r}
fasta_path = file.path(
  params$data_folder,
  paste(params$gene_symbol, "aa_uniprot_fish.fa", sep="_")
)

fish_up_seqs <- extract_up_seqs(protein_features, fasta_path)

fish_up_seqs
```

```{r}
alignment_fn <- file.path(
  params$data_folder,
  paste(params$gene_symbol, "msa_uniprot_fish.fa", sep="_")
)
fish_alignment <- align_isoforms(fish_up_seqs, alignment_fn)

fish_alignment
```

```{r}
aligned_ranges <- convert_alignment_to_ranges(fish_alignment)
msa_position_map <- map_aligned_positions(aligned_ranges)

plot_alignment_models(
  fish_symbol, aligned_ranges,
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "uniprot_proteins_fish.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "PTM", msa_position_map, paste("fish", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "ptms_fish.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "DOMAINS_AND_SITES", msa_position_map, paste("fish", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "domains_fish.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "STRUCTURAL", msa_position_map, paste("fish", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "structure_fish.pdf", sep="_")
  ), use_type=TRUE
)
```

## Drosophila

### Transcript annotations for gene

```{r}
# Mapping of this ID needs to be automated; using hardcoded value for now !!!
fly_symbol <- params$fly_symbol

Tx_d <- transcriptsBy(dmtx, "gene")[[fly_symbol]]
txd_df <- Tx_d %>%
  as.data.frame() %>%
  select(tx_id, tx_name, chromosome=seqnames, start, end)

datatable(txd_df)
```

```{r}
fig_path = file.path(
  params$figure_folder,
  paste(params$gene_symbol, "transcripts_fly.pdf", sep="_")
)
plot_trancript_models(fly_symbol, Tx_d, dmtx, fig_path)
```

### Nucleotide sequences

```{r}
fasta_path = file.path(
  params$data_folder,
  paste(params$gene_symbol, "nt_fly.fa", sep="_")
)
nt_seqs <- dna_seq_from_txdb(Tx_d, dmdb, fasta_path)
nt_seqs
```

### Protein sequences

```{r warning=FALSE, message=FALSE}
protein_features <- fly_symbol %>%
  entries_from_uniprot(id_code="FLYBASE_ID")

protein_features %>%
  select(-SEQ) %>%
  datatable()
```

```{r}
fasta_path = file.path(
  params$data_folder,
  paste(params$gene_symbol, "aa_uniprot_fly.fa", sep="_")
)

fly_up_seqs <- extract_up_seqs(protein_features, fasta_path)

fly_up_seqs
```

```{r}
alignment_fn <- file.path(
  params$data_folder,
  paste(params$gene_symbol, "msa_uniprot_fly.fa", sep="_")
)
fly_alignment <- align_isoforms(fly_up_seqs, alignment_fn)

fly_alignment
```

```{r}
aligned_ranges <- convert_alignment_to_ranges(fly_alignment)
msa_position_map <- map_aligned_positions(aligned_ranges)

plot_alignment_models(
  fly_symbol, aligned_ranges,
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "uniprot_proteins_fly.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "PTM", msa_position_map, paste("fly", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "ptms_fly.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "DOMAINS_AND_SITES", msa_position_map, paste("fly", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "domains_fly.pdf", sep="_")
  )
)
```

```{r message=FALSE, warning=FALSE}
plot_features_on_alignment(
  protein_features, "STRUCTURAL", msa_position_map, paste("fly", params$gene_symbol), 
  file.path(
    params$figure_folder,
    paste(params$gene_symbol, "structure_fly.pdf", sep="_")
  ), use_type=TRUE
)
```